File: cracking_the_coding_interview/2-3.cpp
#include <iostream>
using namespace std;

// 3分16秒 +α (間違ってた)

struct Node{
    int val;
    Node* next;
};

void deleteMiddleNode(Node* node){
    Node* delNode = node->next;
    *node = *node->next; // *(node -> next)
    delete delNode;
}

int main(void){
    Node* node0 = new Node;
    node0->val = 0;
    Node* node1 = new Node;
    node0->next = node1;
    node1->val = 1;
    Node* node2 = new Node;
    node1->next =  node2;
    node2->val = 2;
    cout << node0->next->val << endl;
    deleteMiddleNode(node1);
    cout << node0->next->val << endl;
    return 0;
}

File: cracking_the_coding_interview/2-3.py
class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next

def deleteNode(node):
    del_node = node.next
    node.val = node.next.val
    node.next = node.next.next
    del del_node
    return

def createLinkedList(nums, i):
    if i >= len(nums):
        return None
    
    node = Node(nums[i])
    node.next = createLinkedList(nums, i + 1)
    return node

def output(node):
    while node:
        print(node.val)
        node = node.next

if __name__ == '__main__':
    nums = [0, 1, 2, 3, 4, 5]
    root = createLinkedList(nums, 0)
    deleteNode(root.next.next)
    output(root)